flamingpy/noise/cv.py

```diff
@@ -0,0 +1,623 @@
+        """Return the symplectic matrix associated with CZ application.
+
+        Returns:
+            array: the symplectic matrix.
+        """
+        adj = self._adj
+        return SCZ_mat(adj, sparse)
+
+    # Visualization functions
+    def draw(self, **kwargs):
+        """Draw the CV graph state with matplotlib.
+
+        See flamingpy.utils.viz.draw_EGraph for more details. Use the default
+        colours: gold for GKP states and blue for p-squeezed
+        states.
+        """
+        cv_opts = {"color_nodes": ("state", {"GKP": "gold", "p": "blue"})}
+        updated_opts = {**cv_opts, **kwargs}
+        return self.egraph.draw(**updated_opts)
+
+    def draw_SCZ(self, **kwargs):
+        """Draw the adjacency matrix of a CV graph state with matplotlib.
+
+        See flamingpy.utils.viz.plot_mat_heat_map for more details.
+        """
+        from flamingpy.utils.viz import plot_mat_heat_map
+
+        return plot_mat_heat_map(self.SCZ(), **kwargs)
+
+    # Getters and properties
+    def hom_outcomes(self, inds=None, quad="p"):
+        """array: quad-homodyne measurement outcomes for modes inds."""
+        if inds is None:
+            inds = range(self._N)
+        outcomes = [self.egraph.nodes[self._to_points[i]].get("hom_val_" + quad) for i in inds]
+        return outcomes
+
+    def bit_values(self, inds=None):
+        """array: bit values associated with the p measurement."""
+        if inds is None:
+            inds = range(self._N)
+        bits = [self.egraph.nodes[self._to_points[i]].get("bit_val") for i in inds]
+        return bits
+
+    @property
+    def p_inds(self):
+        """array: the indices of the p-squeezed states."""
+        return self.states.get("p")
+
+    @property
+    def GKP_inds(self):
+        """array: the indices of the GKP states."""
+        return self.states.get("GKP")
+
+    # Private methods
+    def _apply_state_labels(self):
+        """Add state labels as node attributes to the graph state."""
+        for psi in self.states:
+            for ind in self.states[psi]:
+                self.egraph.nodes[self._to_points[ind]]["state"] = psi
+
+    def _covs_sampler(self, sampling_order="initial", inds=None, rng=default_rng()):
+        """Return the covariances for the homodyne measurement sample."""
+        delta = self.delta
+        covs = np.zeros(2 * self._N, dtype=np.float32)
+        if sampling_order == "initial":
+            noise_q = {"p": 1 / (2 * delta) ** 0.5, "GKP": (delta / 2) ** 0.5}
+            noise_p = {"p": (delta / 2) ** 0.5, "GKP": (delta / 2) ** 0.5}
+            for state, inds in self.states.items():
+                if self._perfect_inds:
+                    inds = np.array(list(set(inds).difference(self._perfect_inds)))
+                if len(inds) > 0:
+                    covs[inds] = noise_q[state]
+                    covs[inds + self._N] = noise_p[state]
+            self._init_covs = covs
+
+        elif sampling_order == "two-step":
```

Comment:
It appears current pytest blocks are not covering `sampling_order == "two-step"` cases. This is needed.

Reply:
The test was there (and now in `test_grn_model_two_step`) but due to a bug the block was not actually run. Fixed now - thanks!